src/staking/index.ts

```diff
@@ -1,754 +1,447 @@
+  stakingTxHashHex: string;
+  stakerPkNoCoordHex: string;
+  finalityProviderPkNoCoordHex: string;
+  stakingTx: Transaction;
+  stakingOutputIndex: number;
+  startHeight: number;
+  timelock: number;
 }
 
-// withdrawalTransaction generates a transaction that
-// spends the staking output of the staking transaction
-function withdrawalTransaction(
-  scripts: {
-    timelockScript: Buffer;
-  },
-  scriptTree: Taptree,
-  tx: Transaction,
-  withdrawalAddress: string,
-  network: networks.Network,
-  feeRate: number,
-  outputIndex: number = 0,
-): PsbtTransactionResult {
-  // Check that withdrawal feeRate is bigger than 0
-  if (feeRate <= 0) {
-    throw new Error("Withdrawal feeRate must be bigger than 0");
-  }
-
-  // Check that outputIndex is bigger or equal to 0
-  if (outputIndex < 0) {
-    throw new Error("Output index must be bigger or equal to 0");
-  }
-
-  // position of time in the timelock script
-  const timePosition = 2;
-  const decompiled = script.decompile(scripts.timelockScript);
+export class Staking {
+  network: networks.Network;
+  stakerInfo: StakerInfo;
 
-  if (!decompiled) {
-    throw new Error("Timelock script is not valid");
+  constructor(network: networks.Network, stakerInfo: StakerInfo) {
+    if (!isValidBitcoinAddress(stakerInfo.address, network)) {
+      throw new StakingError(
+        StakingErrorCode.INVALID_INPUT, "Invalid staker bitcoin address",
+      );
+    }
+    if (!isValidNoCoordPublicKey(stakerInfo.publicKeyNoCoordHex)) {
+      throw new StakingError(
+        StakingErrorCode.INVALID_INPUT, "Invalid staker public key",
+      );
+    }
+    this.network = network;
+    this.stakerInfo = stakerInfo;
   }
 
-  let timelock = 0;
+  /**
+   * buildScripts builds the staking scripts for the staking transaction.
+   * Note: different staking types may have different scripts.
+   * e.g the observable staking script has a data embed script.
+   * @param {StakingParams} params - The staking parameters.
+   * @param {string} finalityProviderPkNoCoordHex - The finality provider's public key
+   * without coordinates.
+   * @param {number} timelock - The staking time in blocks.
+   * @param {string} stakerPkNoCoordHex - The staker's public key without coordinates.
+   * 
+   * @returns {StakingScripts} - The staking scripts.
+   */
+  buildScripts(
+    params: StakingParams,
+    finalityProviderPkNoCoordHex: string,
+    timelock: number,
+    stakerPkNoCoordHex: string
+  ): StakingScripts {
+    // Create staking script data
+    let stakingScriptData;
+    try {
+      stakingScriptData = new StakingScriptData(
+        Buffer.from(stakerPkNoCoordHex, "hex"),
+        [Buffer.from(finalityProviderPkNoCoordHex, "hex")],
+        toBuffers(params.covenantNoCoordPks),
+        params.covenantQuorum,
+        timelock,
+        params.unbondingTime
+      );
+    } catch (error: unknown) {
+      throw StakingError.fromUnknown(
+        error, StakingErrorCode.SCRIPT_FAILURE, 
```

Comment:
Same here and across other files: https://i.imgur.com/UJmNuB5.png
I suggest formatting the codebase